util/tls/tls_socket.h

```diff
@@ -114,6 +114,80 @@ class TlsSocket final : public FiberSocketBase {
   std::unique_ptr<Engine> engine_;
   size_t upstream_write_ = 0;
 
+  struct AsyncReqBase {
+    AsyncReqBase(TlsSocket* owner, io::AsyncProgressCb caller_cb, const iovec* vec, uint32_t len)
+        : owner(owner), caller_completion_cb(std::move(caller_cb)), vec(vec), len(len) {
+    }
+
+    TlsSocket* owner;
+    // Callback passed from the user.
+    io::AsyncProgressCb caller_completion_cb;
+
+    const iovec* vec;
+    uint32_t len;
+
+    std::function<void()> continuation;
+    iovec scratch_iovec;
+
+    // Asynchronous helpers
+    void MaybeSendOutputAsyncWithRead();
+
+    // Returns true if we did not start an upstream write
+    void MaybeSendOutputAsync();
+
+    void HandleUpstreamAsyncWrite(io::Result<size_t> write_result, Engine::Buffer buffer);
+
+    void HandleOpAsync(int op_val);
+
+    void StartUpstreamWrite();
+    void StartUpstreamRead();
+
+    void RunPending();
+
+    virtual void Run() = 0;
+    virtual void CompleteAsyncReq(io::Result<size_t> result) = 0;
+  };
+
+  // Helper function that resets the internal async request, applies the
+  // user AsyncProgressCb and returns. We need this, because progress callbacks
+  // can start another async request and for that to work, we need to clean up
+  // the one we are running on.
+  void CompleteAsyncRequest(io::Result<size_t> result);
+
+  struct AsyncWriteReq : AsyncReqBase {
+    using AsyncReqBase::AsyncReqBase;
+
+    // TODO handle async yields to avoid deadlocks (see HandleOp)
+    enum State { PushToEngine, HandleOpAsyncTag, MaybeSendOutputAsyncTag, Done };
+    State state = PushToEngine;
+    PushResult last_push;
+
+    // Main loop
+    void Run() override;
+    void CompleteAsyncReq(io::Result<size_t> result) override;
+  };
+
+  friend AsyncWriteReq;
+
+  struct AsyncReadReq : AsyncReqBase {
+    using AsyncReqBase::AsyncReqBase;
+
+    Engine::MutableBuffer dest;
+    size_t read_total = 0;
+
+    // Main loop
+    void Run() override;
+    void CompleteAsyncReq(io::Result<size_t> result) override;
+  };
+
+  friend AsyncReadReq;
+
+  // TODO clean up the optional before we yield such that progress callback can dispatch another
+  // async operation
+  std::optional<AsyncWriteReq> async_write_req_;
```

Comment:
most of our use with tls is synchronous, so I suggest to make them at least unique_ptr (could even be plain pointers), to reduce the size of TlsSocket for most cases.

this will also allow to move Async state definitions to cc file.

Reply:
Yep that makes sense. It's 130 bytes roughly extra per object